Gems/Terrain/Code/Source/TerrainRaycast/TerrainRaycastContext.cpp

```diff
@@ -77,158 +77,167 @@ namespace
+            clippedLineSegment, AZ::Vector2::CreateZero(), AZ::Vector2(AZStd::numeric_limits<float>::max()), tFromMinCorner);
+
+        // If we're moving in the positive direction in the square, then the amount till the next boundary is actually
+        // the distance remaining to the max corner, not the distance in from the min corner, so flip our calculation.
+        tUntilNextBoundary = AZ::Vector2::CreateSelectCmpGreater(
+            clippedEnd, clippedStart, tDelta - tUntilNextBoundary, tUntilNextBoundary);
+
+        // Initialize our segment/triangle hit tester with the ray that we're using. We use the full ray instead of the clipped one
+        // to make sure we don't run into any precision issues caused from the clipping.
+        AZ::Intersect::SegmentTriangleHitTester hitTester(rayStart, rayEnd);
+
+        // These will hold our current square coordinates in world space values as we loop through the squares, starting with the
+        // grid square for (x0, y0). These values represent the minimum corner of each terrain square.
+        AZ::Vector2 curGridCorner = clippedStartGridCorner;
+
+        // This is how much we need to increment our x and y by to get to the next grid square along the line.
+        // They will either be +/- terrainResolution or 0 if we're not moving in that direction.
+        const AZ::Vector2 gridIncrement = terrainResolution *
+            AZ::Vector2::CreateSelectCmpEqual(
+                                        clippedLineSegment,
+                                        AZ::Vector2::CreateZero(),
+                                        AZ::Vector2::CreateZero(),
+                                        AZ::Vector2(AZ::GetSign(clippedLineSegment.GetX()), AZ::GetSign(clippedLineSegment.GetY())));
+
+        // Convenience vectors that we can use in the loop to just increment one direction.
+        const AZ::Vector2 tDeltaX(tDelta.GetX(), 0.0f);
+        const AZ::Vector2 tDeltaY(0.0f, tDelta.GetY());
+        const AZ::Vector2 gridIncrementX(gridIncrement.GetX(), 0.0f);
+        const AZ::Vector2 gridIncrementY(0.0f, gridIncrement.GetY());
+
+        // Walk through each grid square in the terrain that intersects the XY coordinates of the line.
+        // We'll check each square to see if the ray intersections actually intersect the terrain triangles in the square.
+        for (int terrainSquare = 0; terrainSquare < numTerrainSquares; terrainSquare++)
         {
+            // Create a bounding volume for this terrain square.
+            AZ::Aabb currentVoxel = AZ::Aabb::CreateFromMinMax(
+                AZ::Vector3(curGridCorner, terrainWorldBounds.GetMin().GetZ()),
+                AZ::Vector3(curGridCorner + terrainResolution, terrainWorldBounds.GetMax().GetZ()));
+
+            // Check for a hit against the terrain triangles in this square.
             TriangulateAndFindNearestIntersection(terrainSystem, currentVoxel, hitTester, result);
```

Comment:
1 - I like the idea in concept, but I'm not sure it would be any (or much) cheaper in practice. Assuming the segment hits the terrain *somewhere*, then the min/max heights of the entire segment *should* surround most voxel boxes. So then we'd have to calculate just the min/max heights of the portion of the segment that intersects with the box, which then effectively trades "ray vs triangle" with "ray vs AABB". If you look at the code in IntersectRayAABB vs SegmentTriangleHitTester::Intersect, it's a bit hard to say which one would be faster.

2 - Really good point, this would definitely make the whole thing 50% faster. I'll save this one for a future improvement, but I'll at least add a comment in the code describing this as a future optimization so that the idea is captured.

Reply:
We chatted about this in person, but I wanted to give anyone else that comes across this PR some context. For the first suggestion, what I meant is to get the min/max height values for the ray inside a single terrain quad and compare them with the min/max height values for the 4 samples on the corners of that quad. In the vast majority of cases the ray will be above or below that quad completely, so a triangle test is unnecessary.